masschallenge/django-accelerator
# AC-6662: 500 error received when attempting to delete a user

#### Changes introduced in [AC-6662](https://masschallenge.atlassian.net/browse/AC-6662)
- stop logic from auto creating profiles on query set `get`, instead do that on user creation
- add migration that revokes user deletion permissions from non superuser users

#### How to test
- with a clean db
- while on development on both accelerate and django-accelerator
- get into the **accelerate** django-shell, i.e. `make django-shell`
- perform the following magic that will show you the count of superusers and non-superusers with user deletion powers
```
from django.contrib.auth.models import Permission
from mc.models import User
perm = Permission.objects.filter(name="Can delete user").first()
User.objects.filter(is_superuser=True).count() # take note of this
User.objects.filter(user_permissions=perm).exclude(is_superuser=True).count() # take note of this
```

- [mask](http://localhost:8181/admin/simpleuser/user/4697/masquerade/) as a user that is not a superuser and has user deletion permission
- try and delete [30-user](http://localhost:8181/admin/simpleuser/user/30/delete/), [50193-user](http://localhost:8181/admin/simpleuser/user/50193/delete/) and [38298-user](http://localhost:8181/admin/simpleuser/user/38298/delete/), notice that both of these fail.
   - 50193-user has PartnerTeamMembers  associated
    - 38298-user has Startup Team Members and StartupStatus associated
    - 30-user has both PartnerTeamMembers and Startup Team Members associated
- now checkout AC-6662 on both accelerate and django-accelerator
- run migrations i.e. `make migrate`
- access the **accelerate** django-shell and type in the following
```
from django.contrib.auth.models import Permission
from mc.models import User
perm = Permission.objects.filter(name=“Can delete user").first()
User.objects.filter(is_superuser=True).count() # should be the same as before
User.objects.filter(user_permissions=perm).exclude(is_superuser=True).count() # should be 0
```

- try and delete [30-user](http://localhost:8181/admin/simpleuser/user/30/delete/) and notice you don't have the **delete** button of the page since your delete permissions have been revoked.
- release this user and be demoadmin
- try and delete [30-user](http://localhost:8181/admin/simpleuser/user/30/delete/), [50193-user](http://localhost:8181/admin/simpleuser/user/50193/delete/) and [38298-user](http://localhost:8181/admin/simpleuser/user/38298/delete/) again, notice that it now passes. Success🏆

#### Note
This PR has a sibling on [accelerate](https://github.com/masschallenge/accelerate/pull/2182) and this should be merged in first


Was convinced by @jonkiparsky that the extent of the cleanup initially was necessary but premature. This should be an easier review :)
